ofrak_core/ofrak/core/free_space.py

```diff
@@ -131,19 +135,36 @@ async def allocate_bom(
         # Allocate largest segments first
         segments_to_allocate.sort(key=lambda o_s: o_s[1].length, reverse=True)
         segments_by_object: Dict[str, List[Segment]] = defaultdict(list)
+        possible_perms = None
         for obj, segment in segments_to_allocate:
             vaddr, final_size = 0, 0
             if segment.length > 0:
-                allocs = await self.allocate(
-                    segment.access_perms,
-                    segment.length,
-                    min_fragment_size=segment.length,
-                    alignment=bom.segment_alignment,
-                )
-                allocation = next(iter(allocs))
-                vaddr = allocation.start
-                final_size = allocation.length()
-
+                if permission_map is not None:
+                    possible_perms = permission_map[segment.access_perms]  # type: ignore
+                else:
+                    possible_perms = (segment.access_perms,)
+                for candidate_permissions in possible_perms:
+                    try:
+                        allocs = await self.allocate(
+                            candidate_permissions,
+                            segment.length,
+                            min_fragment_size=segment.length,
+                            alignment=bom.segment_alignment,
+                        )
+                        allocation = next(iter(allocs))
+                        vaddr = allocation.start
+                        final_size = allocation.length()
+                        break
+                    except FreeSpaceAllocationError:
+                        continue
+            if vaddr == 0 or final_size == 0:
+                if possible_perms:
+                    raise FreeSpaceAllocationError(
+                        f"Could not find enough free space for access perms {possible_perms} and "
+                        f"length {segment.length}"
+                    )
+                else:
+                    raise FreeSpaceAllocationError(f"No provided segments longer than 0 bytes.")
```

Comment:
The previous implementation allowed for creating segments with vaddr of 0 and final_size of 0. To preserve this, this entire block should be indented so it runs only if `segment.length > 0`.

Reply:
All of the relevant changes create the following diff:

```
diff --git a/ofrak_core/ofrak/core/free_space.py b/ofrak_core/ofrak/core/free_space.py
index ab4150a..2bae549 100644
--- a/ofrak_core/ofrak/core/free_space.py
+++ b/ofrak_core/ofrak/core/free_space.py
@@ -135,12 +135,11 @@ class Allocatable(ResourceView):
         # Allocate largest segments first
         segments_to_allocate.sort(key=lambda o_s: o_s[1].length, reverse=True)
         segments_by_object: Dict[str, List[Segment]] = defaultdict(list)
-        possible_perms = None
         for obj, segment in segments_to_allocate:
             vaddr, final_size = 0, 0
             if segment.length > 0:
                 if permission_map is not None:
-                    possible_perms = permission_map[segment.access_perms]  # type: ignore
+                    possible_perms = permission_map[segment.access_perms]
                 else:
                     possible_perms = (segment.access_perms,)
                 for candidate_permissions in possible_perms:
@@ -157,14 +156,14 @@ class Allocatable(ResourceView):
                         break
                     except FreeSpaceAllocationError:
                         continue
-            if vaddr == 0 or final_size == 0:
-                if possible_perms:
-                    raise FreeSpaceAllocationError(
-                        f"Could not find enough free space for access perms {possible_perms} and "
-                        f"length {segment.length}"
-                    )
-                else:
-                    raise FreeSpaceAllocationError(f"No provided segments longer than 0 bytes.")
+                if vaddr == 0 or final_size == 0:
+                    if possible_perms:
+                        raise FreeSpaceAllocationError(
+                            f"Could not find enough free space for access perms {possible_perms} and "
+                            f"length {segment.length}"
+                        )
+                    else:
+                        raise FreeSpaceAllocationError(f"No provided segments longer than 0 bytes.")
             segments_by_object[obj.path].append(
                 Segment(
                     segment_name=segment.segment_name,
```